### PR TITLE
`state init` and `state install` supports wildcard notation for language and package versions.

### PR DESCRIPTION
**Pull request by mitchell-as**
<table>
<td>
  <a href="https://activestatef.atlassian.net/browse/DX-2011" title="DX-2011" target="_blank">StoryDX-2011</a>  We can install requirements with wildcard version matching
  </td></table>
  <br />

### CODE REVIEW

File: internal/runners/initialize/init.go [hunk 1]
```diff
@@ -258,6 +258,8 @@ func verifyLangAndVersion(lang language.Language, version string) error {
 		return locale.NewInputError("err_init_language_not_found", "The selected language cannot be found")
 	}
 
+	version = strings.Replace(strings.Replace(version, ".x", "", 1), ".X", "", 1) // strip any wildcard
```
Comment (mitchell-as) [inline internal/runners/initialize/init.go:261]: We're just verifying the wildcarded version exists in the first place. e.g. 3.10.x will check that 3.10 is a valid version prefix.

File: pkg/platform/model/vcs.go [hunk 2]
```diff
@@ -526,29 +512,6 @@ func DeleteBranch(branchID strfmt.UUID) error {
 	return nil
 }
 
-// CommitPackage commits a package to an existing parent commit
```
Comment (mitchell-as) [inline pkg/platform/model/vcs.go:529]: This function is unused.

File: pkg/platform/model/vcs.go [hunk 3]
```diff
@@ -691,23 +676,6 @@ func (cs indexedCommits) countBetween(first, last string) (int, error) {
 	return ct, nil
 }
 
-// CommitRequirement commits a single requirement to the platform
```
Comment (mitchell-as) [inline pkg/platform/model/vcs.go:694]: This function is unused.

File: pkg/platform/model/vcs.go [hunk 4]
```diff
@@ -643,6 +600,34 @@ func CommitInitial(hostPlatform string, langName, langVersion string) (strfmt.UU
 	return res.Payload.CommitID, nil
 }
 
+func versionStringToConstraints(version string) ([]*mono_models.Constraint, error) {
+	constraints := []*mono_models.Constraint{}
```
Comment (mitchell-as) [inline pkg/platform/model/vcs.go:604]: Note that exact versions (no wildcards) will be supported in a separate PR that addresses DX-2010.

File: internal/runners/initialize/init.go [hunk 5]
```diff
@@ -239,60 +238,83 @@ func (r *Initialize) Run(params *RunParams) (rerr error) {
...
 	if err != nil {
-		multilog.Error("Failed to fetch supported languages (using hardcoded default version): %s", errs.JoinMessage(err))
-		return lang.RecommendedVersion()
+		return "", errs.Wrap(err, "Unable to get known versions for language %s", lang.Requirement())
 	}
 
-	for _, l := range langs {
-		if lang.String() == l.Name || (lang == language.Python3 && l.Name == language.Python3.Requirement()) {
-			return l.DefaultVersion
+	prefix := strings.Replace(strings.Replace(version, ".x", "", 1), ".X", "", 1) // strip any wildcard
```
Comment (mitchell-as) [inline internal/runners/initialize/init.go:291]: We're just verifying the wildcarded version exists in the first place. e.g. 3.10.x will check that 3.10 is a valid version prefix.

**Review by daved — CHANGES_REQUESTED** [review]

**Review by daved — APPROVED** [review]

File: internal/runbits/requirements/requirements.go [hunk 6]
```diff
@@ -406,3 +411,19 @@ func initializeProject() (*project.Project, error) {
 
 	return project.FromPath(target)
 }
+
+func versionStringToRequirements(version string) ([]bpModel.VersionRequirement, error) {
+	constraints, err := model.VersionStringToConstraints(version)
+	if err != nil {
+		return nil, errs.Wrap(err, "Unable to process version string into constraints")
+	}
```
Comment (Naatan) [inline internal/runbits/requirements/requirements.go:419]: Can we make this only act if we find that the value isn't just decimal numbers? Most of the time we shouldn't need to reach out to this API.

File: internal/runbits/requirements/requirements.go [hunk 7]
```diff
@@ -406,3 +411,19 @@ func initializeProject() (*project.Project, error) {
...
+
+	requirements := make([]bpModel.VersionRequirement, len(constraints))
+	for i, constraint := range constraints {
+		requirements[i] = bpModel.VersionRequirement{
+			bpModel.VersionRequirementComparatorKey: constraint.Comparator,
+			bpModel.VersionRequirementVersionKey:    constraint.Version,
+		}
+	}
+	return requirements, nil
```
Comment (Naatan) [inline internal/runbits/requirements/requirements.go:428]: I don't like that we're modeling this data outside of the model package. It feels like this should be owned by the model. We should use and enrich the params that model.StageCommit uses.

File: pkg/platform/model/vcs.go [hunk 8]
```diff
@@ -643,6 +600,38 @@ func CommitInitial(hostPlatform string, langName, langVersion string) (strfmt.UU
 	return res.Payload.CommitID, nil
 }
 
+func isWildcardVersion(version string) bool {
+	return strings.Index(version, "x") >= 0 || strings.Index(version, "X") >= 0
```
Comment (Naatan) [inline pkg/platform/model/vcs.go:604]: For `init` you require `x` to be after a dot. Here you don't. I think requiring the dot probably makes more sense, suggests we keep it consistent.

File: pkg/platform/model/vcs.go [hunk 9]
```diff
@@ -643,6 +600,38 @@ func CommitInitial(hostPlatform string, langName, langVersion string) (strfmt.UU
...
+			return nil, locale.NewInputError("err_version_wildcard_start", "A version number cannot start with a wildcard")
+		}
+		constraints = append(constraints, &mono_models.Constraint{Comparator: "gte", Version: strings.Join(parts[:i], ".")})
+		previousPart, err := strconv.Atoi(parts[i-1])
+		if err != nil {
+			return nil, locale.WrapInputError(err, "err_version_number_expected", "Version parts are expected to be numeric")
+		}
+		parts[i-1] = strconv.Itoa(previousPart + 1)
+		constraints = append(constraints, &mono_models.Constraint{Comparator: "lt", Version: strings.Join(parts[:i], ".")})
+	}
```
Comment (Naatan) [inline pkg/platform/model/vcs.go:631]: This does not feel like it should be using mono model structs since we're not sending this to the mono API anymore. Probably makes more sense in the buildplan model?

**Review by Naatan — CHANGES_REQUESTED** [review]

File: internal/runbits/requirements/requirements.go (hunk 6 above)
Comment (mitchell-as) [reply internal/runbits/requirements/requirements.go:419]: This is done in a separate PR: https://github.com/ActiveState/cli/pull/2663/files#diff-d9612958a129acac4c7a4c03e241b6b912c286635b188a24975afe794fcc94f5R605-R618

File: pkg/platform/model/vcs.go (hunk 9 above)
Comment (mitchell-as) [reply pkg/platform/model/vcs.go:631]: We are using the mono API for `state init` (vcs.go's `CommitInitial()` function). In order to avoid duplication, I've decided to have the string -> constraints list translation happen here in vcs.go. BuildPlanner makes use of this, but translates from mono API constraints to build planner constraints.

Let me know if you'd like me to reverse this. The `state init` work came first, so it has precedence in my mind.

File: pkg/platform/model/vcs.go [hunk 10]
```diff
@@ -600,34 +601,18 @@ func CommitInitial(hostPlatform string, langName, langVersion string) (strfmt.UU
...
 }
 
-func isWildcardVersion(version string) bool {
-	return strings.Index(version, "x") >= 0 || strings.Index(version, "X") >= 0
-}
-
-func VersionStringToConstraints(version string) ([]*mono_models.Constraint, error) {
-	if !isWildcardVersion(version) {
-		return []*mono_models.Constraint{{Comparator: "eq", Version: version}}, nil // exact version
+func versionStringToConstraints(version string) ([]*mono_models.Constraint, error) {
```
Comment (Naatan) [inline pkg/platform/model/vcs.go:604]: Not asking for a change, but I just realized I was wrong in the first review in that we don't use the mono model. Filed a followup here: https://activestatef.atlassian.net/browse/DX-2064

**Review by Naatan — APPROVED** [review]
